File: client/src/lib/mindMap.ts
export type MindMapScope = "current_page" | "selected_pages" | "whole_lecture";
export type MindMapDepth = "overview" | "normal" | "detailed";
export interface MindMapNode { id: string; title: string; summary: string; pageReferences: number[]; children: MindMapNode[]; }
export interface MindMapCacheInput { documentId: string; scope: MindMapScope; depth: MindMapDepth; pages: number[]; content: string; }

const asRecord = (value: unknown): Record<string, unknown> => {
  if (!value || typeof value !== "object" || Array.isArray(value)) throw new Error("AI trả về sơ đồ không hợp lệ.");
  return value as Record<string, unknown>;
};
export function parseMindMapResponse(value: unknown): MindMapNode {
  const record = asRecord(asRecord(value).mind_map ?? value);
  const id = record.id, title = record.title, summary = record.summary;
  const pages = record.page_references ?? record.pageReferences ?? [];
  if (typeof id !== "string" || !id || typeof title !== "string" || typeof summary !== "string" || !Array.isArray(pages) || !pages.every((p) => Number.isInteger(p) && p > 0) || !Array.isArray(record.children)) throw new Error("AI trả về sơ đồ không đúng cấu trúc JSON.");
  return { id, title, summary, pageReferences: pages as number[], children: record.children.map(parseMindMapResponse) };
}
export function createMindMapCacheKey(input: MindMapCacheInput) {
  let hash = 5381; for (const char of input.content) hash = (hash * 33) ^ char.charCodeAt(0);
  return `vlearn:mind-map:${input.documentId}:${input.scope}:${input.depth}:${input.pages.join(",")}:${hash >>> 0}`;
}

File: server/app/schemas/mind_map.py
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field


class MindMapPage(BaseModel):
    model_config = ConfigDict(extra="forbid")

    page: int = Field(ge=1)
    text: str = Field(min_length=1, max_length=20_000)


class MindMapRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    content: list[MindMapPage] = Field(min_length=1, max_length=60)
    scope: Literal["current_page", "selected_pages", "whole_lecture"]
    depth: Literal["overview", "normal", "detailed"]


class MindMapNode(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str = Field(min_length=1, max_length=100)
    title: str = Field(min_length=1, max_length=200)
    summary: str = Field(max_length=1_000)
    page_references: list[Annotated[int, Field(ge=1)]] = Field(
        default_factory=list,
        max_length=60,
    )
    children: list["MindMapNode"] = Field(default_factory=list, max_length=30)
